packages/shared/lib/utils/http.ts

```diff
@@ -0,0 +1,92 @@
+/* eslint-disable @typescript-eslint/no-unnecessary-type-parameters */
+import { Err, Ok, redactHeaders, redactObjectOrString, redactURL } from '@nangohq/utils';
+
+import { NangoInternalError } from './error.js';
+
+import type { LogContextStateless } from '@nangohq/logs';
+import type { HTTP_METHOD, MessageHTTPRequest, MessageRow } from '@nangohq/types';
+import type { Result } from '@nangohq/utils';
+
+export async function loggedFetch<TBody>(
+    {
+        url,
+        method,
+        headers,
+        body
+    }: {
+        url: URL;
+        method?: HTTP_METHOD;
+        headers?: Record<string, string> | undefined;
+        body?: string | null;
+    },
+    options: {
+        // TODO: maybe put that in logCtx
+        valuesToFilter: string[];
+        logCtx: LogContextStateless;
+        context: MessageRow['context'];
+    }
+): Promise<Result<{ res: Response; body: TBody }>> {
```

Comment:
what about a function with this signature: 
```
export function withHttpLogging({
    logCtx,
    context,
    valuesToFilter
}: {
    logCtx: LogContextStateless;
    context: MessageRow['context'];
    valuesToFilter: string[];
}): { fetch: (url: URL | string, init?: RequestInit) => Promise<Response> } {
    return {
        fetch: async (url, init) => {
            // ...
            const response = await fetch(url, init);
            // ...
            return response;
        }
    };
}

```

so you can use it exactly like fetch with just some prefixing:
```
const response = await withHttpLogging({logCtx, context: 'auth', ...}).fetch(url, options)
```

Reply:
yes the body of the function is the same except it is returning the fetch response so can be used as a drop-in replacement.

Reply:
I don't think it's possible to do a drop-in replacement, native fetch returns .text()/.json() but they can only be used once otherwise it throws. And also in a drop-in replacement it would throw on network error which we don't want (unless I'm rethrowing but that would means a try/catch just for doing nothing)

Do you have a suggestion?

Reply:
good point. Fine with returning a `Result` then. I just feel the name doesn't fully reflect the behavior of the function. It should be called `fetchThatLogsAndUnwrapBodyAndNotThrowOnNetworkError` ;)
I trust you go for the signature you prefer

Reply:
I'll go like this the, since there is no consensus on naming and can be changed later since it's internal